### PR TITLE
feat: support set property

### PR DESCRIPTION
**Pull request by bzp2010**
## checklist
- [x] code style
- [x] key comments

### CODE REVIEW

**Comment by bzp2010** [comment]
## Update
I found a bug in the `proxy-wasm-go-sdk` project that will break our set property functionality. I submitted issue and the associated fix PR, but based on past commit processing and release, it is not possible to give an exact fix time.

**Issue:** https://github.com/tetratelabs/proxy-wasm-go-sdk/issues/234
**PR:** https://github.com/tetratelabs/proxy-wasm-go-sdk/pull/235

The current feature development and test code writing is done, should we wait for the `proxy-wasm-go-sdk` project developers to merge the fixes and release a new version before continuing this PR?

ping @spacewander

**Comment by bzp2010** [comment]
## Update
PR: [https://github.com/tetratelabs/proxy-wasm-go-sdk/pull/235](https://github.com/tetratelabs/proxy-wasm-go-sdk/pull/235) has been merged into the project by the maintainer and is trying to find out when the next release containing this fix will be.

**Comment by bzp2010** [comment]
## Update

The compilation problem was due to the use of `sudo make`, which caused an exception in `GOROOT` environment variable, so tinygo automatically selected Golang version `1.15.15` in actions VM, causing a compilation error, which I have fixed.

File: src/http/ngx_http_wasm_api.c [hunk 1]
```diff
@@ -359,7 +359,128 @@ int32_t
...
+     * lowercase characters of the property key.
+     */
+    key_lowcase = ngx_http_wasm_get_string_buf(r->pool, path_size);
+    hash = ngx_hash_strlow(key_lowcase, (u_char *) key, path_size);
+
+    /*
+     * Get the ngx_http_core_module main configuration block
+     * and search variables from it.
+     */
+    cmcf = ngx_http_get_module_main_conf(r, ngx_http_core_module);
```
Comment (spacewander) [inline src/http/ngx_http_wasm_api.c:397]: We can save lots of code by calling ngx_http_lua_ffi_var_set directly? See https://github.com/openresty/lua-resty-core/blob/master/lib/resty/core/var.lua#L134

Comment (bzp2010) [reply src/http/ngx_http_wasm_api.c:397]: Yes, direct reuse is the best way, I will try it out.

Comment (bzp2010) [reply src/http/ngx_http_wasm_api.c]: changed

**Review by spacewander — COMMENTED** [review]
Let's copy the `proxy_set_property` from the spec to https://github.com/api7/wasm-nginx-module/blob/main/proxy_wasm_abi.md#proxy_get_property

File: src/http/ngx_http_wasm_api.c [hunk 2]
```diff
@@ -359,6 +363,49 @@ int32_t
...
+    if (key == NULL) {
+        return PROXY_RESULT_INVALID_MEMORY_ACCESS;
+    }
+
+    /*
+     * Request a piece of temporary memory to store the
+     * lowercase characters of the property key and errmsg
+     * of variable setting.
+     */
+    key_lowcase = ngx_palloc(r->pool, path_size + errlen);
```
[the remark this answers is not in the document]
Comment (bzp2010) [reply src/http/ngx_http_wasm_api.c:393]: Sorry, I used palloc for debugging purposes and forgot to change it back.

[the remark this answers is not in the document]
Comment (bzp2010) [reply src/http/ngx_http_wasm_api.c:393]: changed

File: src/http/ngx_http_wasm_api.c [hunk 3]
```diff
@@ -359,6 +363,49 @@ int32_t
...
+    ngx_http_request_t          *r;
+
+    log = ngx_http_wasm_get_log();
+    must_get_req(r);
+
+    /* fetch property key and value */
+    key = ngx_wasm_vm.get_memory(log, path_data, path_size);
+    if (key == NULL) {
+        return PROXY_RESULT_INVALID_MEMORY_ACCESS;
+    }
```
[the remark this answers is not in the document]
Comment (bzp2010) [reply src/http/ngx_http_wasm_api.c:382]: changed

File: src/http/ngx_http_wasm_api.c [hunk 4]
```diff
@@ -359,6 +363,49 @@ int32_t
...
+     * Request a piece of temporary memory to store the
+     * lowercase characters of the property key and errmsg
+     * of variable setting.
+     */
+    key_lowcase = ngx_palloc(r->pool, path_size + errlen);
+    errmsg = key_lowcase + path_size;
+
+    /* Call the functions in lua-resty-core to set the variables. */
+    result = set_variable(r, (u_char *) key, path_size, key_lowcase,
+        (u_char *) value, size, errmsg, &errlen);
```
[the remark this answers is not in the document]
Comment (bzp2010) [reply src/http/ngx_http_wasm_api.c:398]: changed

File: src/http/ngx_http_wasm_api.c [hunk 5]
```diff
@@ -359,6 +363,49 @@ int32_t
...
+
+    if (result != NGX_OK) {
+        ngx_log_error(NGX_LOG_ERR, log, 0, (const char *)errmsg);
+
+        if (ngx_strstrn(errmsg, "not found for writing", 20) != NULL) {
+            return PROXY_RESULT_NOT_FOUND;
+        }
```
[the remark this answers is not in the document]
Comment (bzp2010) [reply src/http/ngx_http_wasm_api.c:405]: changed

File: t/property.t [hunk 6]
```diff
@@ -46,3 +46,65 @@ location /t {
...
+        local plugin_ctx, err = wasm.on_configure(plugin, 'for_test|new_value')
+        ngx.say(ngx.var.for_test)
+        ngx.say(#ngx.var.for_test)
+        assert(wasm.on_http_request_headers(plugin_ctx))
+        ngx.say(ngx.var.for_test)
+        ngx.say(#ngx.var.for_test)
+    }
+}
+--- request
+GET /t
```
[the remark this answers is not in the document]
Comment (bzp2010) [reply t/property.t]: cleaned

**Review by spacewander — APPROVED** [review]

File: proxy_wasm_abi.md [hunk 7]
```diff
@@ -268,6 +268,17 @@ Remove key (`key_data`, `key_size`) from a given map (`map_type`).
 
 Get data such as Nginx variables and plugin ID, Use path_data as key and write the value of the obtained variable to res_data.
 
+
+### `proxy_get_property`
```
Comment (spacewander) [inline proxy_wasm_abi.md:272]:
```suggestion
### `proxy_set_property`
```

**Pull request merged**